test/unit/app/controllers/permissions/permissions-controller-test.js

```diff
@@ -0,0 +1,1392 @@
+      assert.deepEqual(
+        permissionsHistory, {},
+        'should not have modified history'
+      )
+      assert.deepEqual(
+        notifications[ORIGINS.a], [],
+        'should not have sent notification'
+      )
+    })
+
+    it('throws if called with bad accounts', async function () {
+
+      assert.rejects(
+        permController.legacyExposeAccounts(ORIGINS.a, []),
+        ERRORS.validatePermittedAccounts.invalidParam(),
+        'should throw if called with no accounts'
+      )
+
+      const permissionsHistory = permController.permissionsLog.getHistory()
+      assert.deepEqual(
+        permissionsHistory, {},
+        'should not have modified history'
+      )
+      assert.deepEqual(
+        notifications[ORIGINS.a], [],
+        'should not have sent notification'
+      )
+    })
+
+    it('throws if called with bad origin', async function () {
+
+      assert.rejects(
+        permController.legacyExposeAccounts(null, ACCOUNT_ARRAYS.a),
+        ERRORS.legacyExposeAccounts.badOrigin(),
+        'should throw if called with invalid origin'
+      )
+
+      const permissionsHistory = permController.permissionsLog.getHistory()
+      assert.deepEqual(
+        permissionsHistory, {},
+        'should not have modified history'
+      )
+      Object.keys(notifications).forEach((domain) => {
+        assert.deepEqual(
+          notifications[domain], [],
+          'should not have sent notification'
+        )
+      })
+    })
+  })
+
+  describe('handleNewAccountSelected', function () {
+
+    let permController, notifications
+
+    beforeEach(function () {
+      notifications = initNotifications()
+      permController = initPermController(notifications)
+      grantPermissions(
+        permController, ORIGINS.a,
+        PERMS.finalizedRequests.eth_accounts(ACCOUNT_ARRAYS.a)
+      )
+      grantPermissions(
+        permController, ORIGINS.b,
+        PERMS.finalizedRequests.eth_accounts(ACCOUNT_ARRAYS.b)
+      )
+    })
+
+    it('throws if invalid origin or account', async function () {
+
+      assert.rejects(
+        permController.handleNewAccountSelected({}, DUMMY_ACCOUNT),
+        ERRORS.handleNewAccountSelected.invalidParams(),
+        'should throw if origin non-string'
+      )
+
+      assert.rejects(
+        permController.handleNewAccountSelected('', DUMMY_ACCOUNT),
+        ERRORS.handleNewAccountSelected.invalidParams(),
+        'should throw if origin empty string'
+      )
+
+      assert.rejects(
+        permController.handleNewAccountSelected(ORIGINS.a, {}),
+        ERRORS.handleNewAccountSelected.invalidParams(),
+        'should throw if account non-string'
+      )
+
+      assert.rejects(
+        permController.handleNewAccountSelected(ORIGINS.a, ''),
+        ERRORS.handleNewAccountSelected.invalidParams(),
+        'should throw if account empty string'
+      )
+    })
+
+    it('does nothing if account not permitted for origin', async function () {
```

Comment:
As discussed elsewhere, this is handled elsewhere, but created tracking issue because we can handle it better: https://github.com/MetaMask/metamask-extension/issues/8187